### ISSUE TITLE
DOC: from dev, version switcher doesn't switch correctly

### ISSUE THREAD

**Issue by mdhaber** [scipy/scipy#18031]
### Issue with current documentation:

In SciPy 1.10.1, we can click `[source]` to visit the source code underlying a feature, and the version switcher can be used to switch to past or future versions of the feature documentation. For example, the documentation of [`scipy.linalg.solve`](https://docs.scipy.org/doc/scipy-1.10.1/reference/generated/scipy.linalg.solve.html#scipy.linalg.solve) looks like:

![image](https://user-images.githubusercontent.com/6570539/220792150-92ff8aec-373d-4672-92ef-1fd192c85560.png)

In dev, however, the `[source]` link no longer appears, and the version switcher navigates to `index.html` of the selected version.

![image](https://user-images.githubusercontent.com/6570539/220792478-76a3f8d4-08ee-470c-b2fa-6545ac73d59e.png)

**Comment by tupui** [comment]
Yes I think I broke it with the conf.py cleaning. There was some inspect thing twice... Will look tmr

**Comment by mdhaber** [comment]
Thanks @tupui. gh-18037 fixed the [`source`] part of this issue, but the version switcher is still misbehaving in dev documentation.

**Comment by tupui** [comment]
Can you clarify this? On my side it's working as expected. (Note that it can also be that some fixes for some browsers got in and we cannot get them due to not being able to update the theme.)

**Comment by mdhaber** [comment]
[Here](http://somup.com/c0nTFDyXTd)'s a 20-second video demonstrating the problem.

**Comment by tupui** [comment]
Indeed I see, I just tested this now and it's only when going from `dev` to another version that it happens.

@choldgraf is this a known issue?

If we are on https://scipy.github.io/devdocs/reference/datasets.html
Then selecting 1.10.1 does not go to https://docs.scipy.org/doc/scipy-1.10.1/reference/datasets.html

The inverse works.

Here is our conf. Not sure why it's an issue, NumPy seem to not have this issue but they are on the same domain. Could that be the issue?
```
    {
        "name": "dev",
        "version": "dev",
        "url": "https://scipy.github.io/devdocs/"
    },
    {
        "name": "1.10.1 (stable)",
        "version":"1.10.1",
        "url": "https://docs.scipy.org/doc/scipy-1.10.1/"
    },
```

**Comment by tylerjereddy** [comment]
bumping milestone, if you need an urgent fix, better merge it soon because I plan to branch May 29th

**Comment by melissawm** [comment]
I *think* this is related to #18535, but still investigating. Currently, when changing from dev to stable on the datasets page, I see the following message on the browser debugger: (sorry can't get it in english but it will make sense):

```
Requisição cross-origin bloqueada: A diretiva Same Origin (mesma origem) não permite a leitura do recurso remoto em https://docs.scipy.org/doc/scipy-1.10.1/reference/datasets.html (motivo: falta cabeçalho 'Access-Control-Allow-Origin' no CORS). Código de status: 200.
```

Then link there points to this document: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSMissingAllowOrigin?utm_source=devtools&utm_medium=firefox-cors-errors&utm_campaign=default

I believe this would explain @tupui's comment above - and the fix would be for someone with access to the server's config to do the following:

> If the server is under your control, add the origin of the requesting site to the set of domains permitted access by adding it to the Access-Control-Allow-Origin header's value.
> 
> For example, to allow a site at https://example.com to access the resource using CORS, the header should be:
> 
> ```
> Access-Control-Allow-Origin: https://example.com
> ```

**Comment by drammock** [comment]
> I am not sure how CORS is handled for the doc. @drammock would you know? Is this a known issue? Can we tweak CORS to allow different dev and prod servers?

sorry for the 4+ month delayed response. The version switcher was buggy and has been fixed in https://github.com/pydata/pydata-sphinx-theme/pull/1395, which made it into version 0.14.0 of the theme. It still won't overcome CORS restrictions though: that must be done in server config I think. So if your dev docs and stable docs are at different TLDs I don't think upgrading the theme to 0.14.0 will fix the problem.

Note that RTD tried to address the CORS issue (for switching between PR builds and deployed builds, which are on different TLDs) with https://github.com/readthedocs/readthedocs.org/pull/10737 but I haven't actually tested whether it works ([this comment](https://github.com/readthedocs/readthedocs.org/pull/10737/files#r1329361936) makes me skeptical whether it will)

**Comment by melissawm** [comment]
Going to https://docs.scipy.org/doc/scipy/, the version switcher now seems to not be working at all. I will try and take a look as soon as possible but if it's a CORS issue not sure what else we can do from this side...

**Comment by tylerjereddy** [comment]
Before you spend too much time, I'm almost certain it is because after merging PRs like this one (https://github.com/scipy/docs.scipy.org/pull/84) we need an extra upload step. Usually Ralf does it but I can do it too, though it probably won't happen today.

I'll make a note to circle back, but you'll probably notice that the version switcher is missing for `1.11.3` and `1.11.4`, but is "ok" for prior versions, and that `1.11.3` and `1.11.4` also aren't listed on docs.scipy.org version list, so that's a likely culprit.

I'll make a calendar note to circle back...

**Comment by lucascolley** [comment]
An update here:

> "In dev ... the version switcher navigates to index.html of the selected version."

This problem is still here.

> 1.11.3 and 1.11.4 also aren't listed on docs.scipy.org version list

This is still the case.

> you'll probably notice that the version switcher is missing for 1.11.3 and 1.11.4

It seems fine for me now on 1.11.3. On 1.11.4 nothing happens when clicking the button.

**Comment by tylerjereddy** [comment]
Yeah, I didn't do what I said I'd do, I'll try to do it today, but prioritizing RC1 for `1.12.0` first.

**Comment by tylerjereddy** [comment]
I just did `make dist` followed by `make upload USER=treddy`, so that solves 1 of the problems:

> 1.11.3 and 1.11.4 also aren't listed on docs.scipy.org version list

Maybe things will get a bit better after cache clearing, but I suspect that didn't actually fix most of the other issues unfortunately. If there are specific requests for changes feel free to ping me directly. I do occasionally forget to adjust version switcher things..

**Comment by lucascolley** [comment]
On Safari iOS for just 1.11.4, the menu button (which should contain the version switcher) is unresponsive, not just the version switcher. Hopefully this isn't too hard to fix as it's just on the one version?

![image](https://github.com/scipy/scipy/assets/51488791/cbe20f9e-28df-40ce-8adf-f8ef6e805515)

**Comment by mdhaber** [comment]
Looks like this is not working again.

**Comment by melissawm** [comment]
Ok - so the version switcher "works", but it always redirects to the landing page for the docs instead of the corresponding page in older versions. I see CORS issues in the console when trying (screenshot in portuguese but hopefully not incomprehensible):

![Captura de imagem_20250110_095743](https://github.com/user-attachments/assets/35fa033d-1fca-4dca-b683-b7550f212550)

**Comment by nickodell** [comment]
> I see CORS issues in the console when trying (screenshot in portuguese but hopefully not incomprehensible):

Indeed, CORS seems to be the cause of the issue.

I used my browser's debugger to break on each XHR request, and it found the following block of code.

```
async function checkPageExistsAndRedirect(event) {
  // ensure we don't follow the initial link
  event.preventDefault();
  const currentFilePath = getCurrentUrlPath();
  let tryUrl = event.currentTarget.getAttribute("href");
  let otherDocsHomepage = tryUrl.replace(currentFilePath, "");
  try {
    let head = await fetch(tryUrl, { method: "HEAD" });
    if (head.ok) {
      location.href = tryUrl; // the page exists, go there
    } else {
      location.href = otherDocsHomepage;
    }
  } catch (err) {
    // something went wrong, probably CORS restriction, fallback to other docs homepage
    location.href = otherDocsHomepage;
  }
}
```

Stepping through this with a debugger, it gets to the line `let head = await fetch(tryUrl, { method: "HEAD" });`. The variable `tryURL` is set to `"https://docs.scipy.org/doc/scipy-1.16.2/reference/datasets.html"`, which is a valid URL.

This fails with the error:

* `NetworkError when attempting to fetch resource.` (Firefox)
* `TypeError: Failed to fetch at HTMLAnchorElement.S` (Chrome)

Because this CORS request fails, it jumps to the line `// something went wrong, probably CORS restriction, fallback to other docs homepage`, and goes to the docs homepage.

Thinking about why this fails, this is not a "simple" request. A "simple" request is one which is a GET, HEAD, or POST, only sets a few allowed headers, and does not attempt to read the result of the request. The docs.scipy.org domain also does not set the Access-Control-Allow-Origin header, so the request fails.

> Indeed I see, I just tested this now and it's only when going from dev to another version that it happens. [...] The inverse works.

It looks like this is because GitHub sets the `Access-Control-Allow-Origin: *` header, which allows all CORS requests going from stable -> dev.

>I believe this would explain @tupui's comment above - and the fix would be for someone with access to the server's config to do the following:

I concur.

I just tried this, using HTTP Toolkit to rewrite the headers on the request in flight, and I find that adding a header of `Access-Control-Allow-Origin: https://scipy.github.io` in response to the HEAD request causes the version switcher to find the right page. I believe we could send this header in response to any HEAD or OPTIONS request and that'd work fine.
